### PR TITLE
static analysis

### PR DESCRIPTION
**Pull request by mrT23**
### **PR Type**
Enhancement, Tests, Documentation, Dependencies

___

### **Description**
- Introduce static analysis tool with multi-language support using tree-sitter.
- Implement file summarization functionality for various programming languages.
- Add comprehensive unit tests for the static analysis components.
- Define and integrate tree-sitter query grammars for multiple languages.
- Update project dependencies to include tree-sitter and related packages.

___

### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><details><summary>17 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>file_summary.py</strong><dd><code>Implement FileSummary class with multi-language support</code>&nbsp; &nbsp; </dd></summary>
<hr>

pr_agent/static_analysis/src/file_summary.py

<li>Implemented <code>FileSummary</code> class for summarizing file content.<br> <li> Added support for multiple programming languages using tree-sitter.<br> <li> Integrated tree-sitter queries for code analysis.<br> <li> Added helper functions for language detection.<br>

</details>

  </td>
  <td><a href="https://github.com/Codium-ai/pr-agent/pull/1266/files#diff-55131e5c06122d14f73664abb45051c7ece6453d3f563d6b29227df862578634">+150/-0</a>&nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>tree-sitter-c-tags.scm</strong><dd><code>Add tree-sitter queries for C</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

pr_agent/static_analysis/queries/tree-sitter-c-tags.scm

<li>Added tree-sitter queries for C language definitions and references.<br>

</details>

  </td>
  <td><a href="https://github.com/Codium-ai/pr-agent/pull/1266/files#diff-b9aeaf67cd4331c234a1fc4e678aa5df97e82602055e788b8dbae8100b4f6e3c">+9/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>tree-sitter-c_sharp-tags.scm</strong><dd><code>Add tree-sitter queries for C#</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

pr_agent/static_analysis/queries/tree-sitter-c_sharp-tags.scm

<li>Added tree-sitter queries for C# language definitions and references.<br>

</details>

  </td>
  <td><a href="https://github.com/Codium-ai/pr-agent/pull/1266/files#diff-6193d6a767712b6d919df4183b066a6413cd8a535ee93a862606b9ec20763697">+46/-0</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>tree-sitter-cpp-tags.scm</strong><dd><code>Add tree-sitter queries for C++</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

pr_agent/static_analysis/queries/tree-sitter-cpp-tags.scm

<li>Added tree-sitter queries for C++ language definitions and references.<br> <br>

</details>

  </td>
  <td><a href="https://github.com/Codium-ai/pr-agent/pull/1266/files#diff-33e8f4f09ce66bcf26ae0a885273f0ddbd2807616b4e76f82ad15dd8ac9ef2d0">+15/-0</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>tree-sitter-elisp-tags.scm</strong><dd><code>Add tree-sitter queries for Emacs Lisp</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

pr_agent/static_analysis/queries/tree-sitter-elisp-tags.scm

<li>Added tree-sitter queries for Emacs Lisp language definitions and <br>references.<br>

</details>

  </td>
  <td><a href="https://github.com/Codium-ai/pr-agent/pull/1266/files#diff-831638c7ee9d9187ecc8c7e321a30d805fad09883c9ddaf949db502116635efd">+8/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>tree-sitter-elixir-tags.scm</strong><dd><code>Add tree-sitter queries for Elixir</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

pr_agent/static_analysis/queries/tree-sitter-elixir-tags.scm

<li>Added tree-sitter queries for Elixir language definitions and <br>references.<br>

</details>

  </td>
  <td><a href="https://github.com/Codium-ai/pr-agent/pull/1266/files#diff-52b20fee3955fd9d1f3a19115d13519178eb30d245fa37dc32e34dfe7b5bf633">+54/-0</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>tree-sitter-elm-tags.scm</strong><dd><code>Add tree-sitter queries for Elm</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

pr_agent/static_analysis/queries/tree-sitter-elm-tags.scm

<li>Added tree-sitter queries for Elm language definitions and references.<br> <br>

</details>

  </td>
  <td><a href="https://github.com/Codium-ai/pr-agent/pull/1266/files#diff-18498e2956a4c953adea402c88810cf4476c6c1e8499acbcd7fd39703757666f">+19/-0</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>tree-sitter-go-tags.scm</strong><dd><code>Add tree-sitter queries for Go</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

pr_agent/static_analysis/queries/tree-sitter-go-tags.scm

<li>Added tree-sitter queries for Go language definitions and references.<br>

</details>

  </td>
  <td><a href="https://github.com/Codium-ai/pr-agent/pull/1266/files#diff-a750eb5c2f0c37b9df11994137bf1b7bc10e27f09c272a0c201a4681febcc73e">+30/-0</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>tree-sitter-java-tags.scm</strong><dd><code>Add tree-sitter queries for Java</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

pr_agent/static_analysis/queries/tree-sitter-java-tags.scm

<li>Added tree-sitter queries for Java language definitions and <br>references.<br>

</details>

  </td>
  <td><a href="https://github.com/Codium-ai/pr-agent/pull/1266/files#diff-ecdaeca73783af17bbb7dcec4c4351bc1405604ff1001eb7916ffc81c56ca6bf">+20/-0</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>tree-sitter-javascript-tags.scm</strong><dd><code>Add tree-sitter queries for JavaScript</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

pr_agent/static_analysis/queries/tree-sitter-javascript-tags.scm

<li>Added tree-sitter queries for JavaScript language definitions and <br>references.<br>

</details>

  </td>
  <td><a href="https://github.com/Codium-ai/pr-agent/pull/1266/files#diff-5e5a2037652325bacd4566306cce07efb23e06c1167426bfd596408e9abafb1f">+88/-0</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>tree-sitter-ocaml-tags.scm</strong><dd><code>Add tree-sitter queries for OCaml</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

pr_agent/static_analysis/queries/tree-sitter-ocaml-tags.scm

<li>Added tree-sitter queries for OCaml language definitions and <br>references.<br>

</details>

  </td>
  <td><a href="https://github.com/Codium-ai/pr-agent/pull/1266/files#diff-fa3057a4257dedfdfbb1f55cb745b1d4212956b96eed7157afc1c8d9d52279a7">+115/-0</a>&nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>tree-sitter-php-tags.scm</strong><dd><code>Add tree-sitter queries for PHP</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

pr_agent/static_analysis/queries/tree-sitter-php-tags.scm

<li>Added tree-sitter queries for PHP language definitions and references.<br> <br>

</details>

  </td>
  <td><a href="https://github.com/Codium-ai/pr-agent/pull/1266/files#diff-3b995d1332fa6c6ed95449edd90acf3f46d079ee77ddb93facfbf1d9ef01d179">+26/-0</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>tree-sitter-python-tags.scm</strong><dd><code>Add tree-sitter queries for Python</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

pr_agent/static_analysis/queries/tree-sitter-python-tags.scm

<li>Added tree-sitter queries for Python language definitions and <br>references.<br>

</details>

  </td>
  <td><a href="https://github.com/Codium-ai/pr-agent/pull/1266/files#diff-62e8e8c2fbe3a17c0a402d5a912c9388c984f0efa46ae94e8e767d795c50ef6d">+12/-0</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>tree-sitter-ql-tags.scm</strong><dd><code>Add tree-sitter queries for QL</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

pr_agent/static_analysis/queries/tree-sitter-ql-tags.scm

<li>Added tree-sitter queries for QL language definitions and references.<br>

</details>

  </td>
  <td><a href="https://github.com/Codium-ai/pr-agent/pull/1266/files#diff-6aee2152fa80176a0f86d391cbfba18fd614d2e1bae632f9b3775dc187e9cdf9">+26/-0</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>tree-sitter-ruby-tags.scm</strong><dd><code>Add tree-sitter queries for Ruby</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

pr_agent/static_analysis/queries/tree-sitter-ruby-tags.scm

<li>Added tree-sitter queries for Ruby language definitions and <br>references.<br>

</details>

  </td>
  <td><a href="https://github.com/Codium-ai/pr-agent/pull/1266/files#diff-b560fe6f038650f7049b229d3e02081212703c9d9bece738c297ee5ac30e07f5">+64/-0</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>tree-sitter-rust-tags.scm</strong><dd><code>Add tree-sitter queries for Rust</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

pr_agent/static_analysis/queries/tree-sitter-rust-tags.scm

<li>Added tree-sitter queries for Rust language definitions and <br>references.<br>

</details>

  </td>
  <td><a href="https://github.com/Codium-ai/pr-agent/pull/1266/files#diff-cf8b00deefb2f4881207cb91a7d152a2c105e30f0ebc90b384c0c33b656c62de">+60/-0</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>tree-sitter-typescript-tags.scm</strong><dd><code>Add tree-sitter queries for TypeScript</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

pr_agent/static_analysis/queries/tree-sitter-typescript-tags.scm

<li>Added tree-sitter queries for TypeScript language definitions and <br>references.<br>

</details>

  </td>
  <td><a href="https://github.com/Codium-ai/pr-agent/pull/1266/files#diff-867df1159e7f40b9aaeb0f722acbe132945b31acdec879582be32cb9aeedead6">+41/-0</a>&nbsp; &nbsp; </td>

</tr>                    
</table></details></td></tr><tr><td><strong>Tests</strong></td><td><details><summary>4 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>test_file_summary.py</strong><dd><code>Add unit tests for FileSummary</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

pr_agent/static_analysis/tests/test_file_summary.py

<li>Added unit tests for <code>FileSummary</code> class.<br> <li> Implemented tests for C++, TypeScript, and Java files.<br>

</details>

  </td>
  <td><a href="https://github.com/Codium-ai/pr-agent/pull/1266/files#diff-0f8611a8b110032994e30466679af4a79b8d3d1e07dd7e44bbdf1b64eb0cf8fd">+43/-0</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>AES.cpp</strong><dd><code>Add AES encryption implementation in C++</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

pr_agent/static_analysis/tests/example_files/AES.cpp

<li>Added AES encryption and decryption implementation.<br> <li> Included optimized code using AESNI and Neon instructions.<br> <li> Integrated with mbedtls library.<br>

</details>

  </td>
  <td><a href="https://github.com/Codium-ai/pr-agent/pull/1266/files#diff-0379e82680ccba589a7139d7a99847a35ccea05c549bafde0fd701fbb7e86d2f">+428/-0</a>&nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>calc.java</strong><dd><code>Add Calculator UI in Java</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

pr_agent/static_analysis/tests/example_files/calc.java

<li>Added Java-based Calculator UI implementation.<br> <li> Implemented GUI components and event handlers.<br>

</details>

  </td>
  <td><a href="https://github.com/Codium-ai/pr-agent/pull/1266/files#diff-8d8b0d35e39ea5db489083ad0f0fc6b4de8cbecb696edc2aba760e5d25f25360">+586/-0</a>&nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>match.ts</strong><dd><code>Add match expression in TypeScript</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

pr_agent/static_analysis/tests/example_files/match.ts

<li>Added TypeScript match expression implementation.<br> <li> Implemented pattern matching utilities.<br>

</details>

  </td>
  <td><a href="https://github.com/Codium-ai/pr-agent/pull/1266/files#diff-e1a6f6480af1dc2af4580ea4c772f7d6e7b8f43f1f3ce05ecb59b6d66bcc4f72">+138/-0</a>&nbsp; </td>

</tr>                    
</table></details></td></tr><tr><td><strong>Documentation</strong></td><td><details><summary>1 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>Credits.md</strong><dd><code>Add Credits documentation</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

pr_agent/static_analysis/queries/Credits.md

- Added credits section acknowledging sources.

</details>

  </td>
  <td><a href="https://github.com/Codium-ai/pr-agent/pull/1266/files#diff-8083298d59e639ff0feb1f865be2e2218e2aa3fe1f390a055af73c7d0a0f5ff9">+5/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    
</table></details></td></tr><tr><td><strong>Dependencies</strong></td><td><details><summary>1 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>requirements.txt</strong><dd><code>Update dependencies for static analysis</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

requirements.txt

<li>Added <code>tree_sitter</code> and related packages.<br> <li> Updated static analysis tool dependencies.<br>

</details>

  </td>
  <td><a href="https://github.com/Codium-ai/pr-agent/pull/1266/files#diff-4d7c51b1efe9043e44439a949dfd92e5827321b34082903477fd04876edb7552">+6/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    
</table></details></td></tr></tr></tbody></table>

___

> 💡 **PR-Agent usage**: Comment `/help "your question"` on any pull request to receive relevant information

### CODE REVIEW

**Review by hussam789 — APPROVED** [review]

**Comment by mrT23** [comment]
Preparing PR description...

**Comment by mrT23** [comment]
Preparing PR description...

**Comment by mrT23** [comment]
**[PR Description](https://github.com/Codium-ai/pr-agent/pull/1266)** updated to latest commit (https://github.com/Codium-ai/pr-agent/commit/88c2b90860af311283ed4dc3e82f1d7b1294e90d)

**Comment by mrT23** [comment]
Preparing PR description...

**Comment by mrT23** [comment]
**[PR Description](https://github.com/Codium-ai/pr-agent/pull/1266)** updated to latest commit (https://github.com/Codium-ai/pr-agent/commit/88c2b90860af311283ed4dc3e82f1d7b1294e90d)

**Comment by mrT23** [comment]
Preparing PR description...

**Comment by mrT23** [comment]
Preparing PR description...

**Comment by mrT23** [comment]
Preparing PR description...

**Comment by mrT23** [comment]
**[PR Description](https://github.com/Codium-ai/pr-agent/pull/1266)** updated to latest commit (https://github.com/Codium-ai/pr-agent/commit/88c2b90860af311283ed4dc3e82f1d7b1294e90d)

**Comment by mrT23** [comment]
## Generating PR code suggestions


Work in progress ...<br>

**Comment by mrT23** [comment]
/help "I am trying to understand how best_practices.md is being used in this document [section](https://qodo-merge-docs.qodo.ai/tools/improve/#best-practices)
Is this relevant_best_practices [here](https://github.com/Codium-ai/pr-agent/blob/e9769fa60263505a00b2ed5e7e566f71cef75318/pr_agent/tools/pr_code_suggestions.py#L75) retrieving info from best_practices.md file and add to the LLM context? I am not seeing where best_practices.md is being used in the code."

File: requirements.txt
Comment (hussam789) [inline requirements.txt:1]: new requirements

**Comment by mrT23** [comment]
/help "how to get more customized code suggestions" ?

**Comment by mrT23** [comment]
/help "how to get more customized code suggestions" ?

**Comment by mrT23** [comment]
/help "how to get more customized code suggestions" ?

**Comment by mrT23** [comment]
## Generating PR code suggestions


Work in progress ...<br>

**Comment by mrT23** [comment]
## Generating PR code suggestions


Work in progress ...<br>

**Comment by mrT23** [comment]
## Generating PR code suggestions


Work in progress ...<br>

**Comment by mrT23** [comment]
## Generating PR code suggestions


Work in progress ...<br>

**Comment by mrT23** [comment]
## Generating PR code suggestions


Work in progress ...<br>

**Comment by mrT23** [comment]
## Generating PR code suggestions


Work in progress ...<br>

**Comment by mrT23** [comment]
## Generating PR code suggestions


Work in progress ...<br>

**Comment by mrT23** [comment]
## Generating PR code suggestions


Work in progress ...<br>

**Pull request closed without merging**